spec/models/redirect_spec.rb

```diff
@@ -0,0 +1,32 @@
+describe Redirect do
+  describe 'validations' do
+    subject do
+      described_class.new(source: 'foo',
+                          destination: 'bar',
+                          redirect_type: 'temporary')
+    end
+
+    it { is_expected.to validate_presence_of(:source)  }
+    it { is_expected.to validate_presence_of(:destination)  }
+    it { is_expected.to validate_presence_of(:redirect_type)  }
+    it { is_expected.to validate_uniqueness_of(:source) }
+
+    it 'validates source and destination not identical' do
+      subject.destination = 'foo'
+      expect(subject).to_not be_valid
```

Comment:
A personal preference, I always prefer to be specific on this specs. For two reasons, it is very difficult to trace where is the problem. If you change the implementation in some way, RSpec won't produce a descriptive error message on the record be_valid. It will simply says "expected valid? to return false, got true." 
Did you try:

``ruby
  expect(subject).to have(1).error_on(:destination)
```

And change the implementation to add the error:

```ruby
  errors.add(:destination, :source_duplicated)
```

And in the locale you create a key called source_duplicated.

```
subject.errors[:source]